README.md
# AviSynth Shader v1.6.6

<a href="https://github.com/mysteryx93/AviSynthShader/releases">Download here >></a>

This plugin allows running HLSL pixel shaders within AviSynth. This gives access to various HLSL filters that haven't been programmed in AviSynth.

by <a href="https://www.spiritualselftransformation.com/">Etienne Charland</a>

This implementation allows running several shaders in a row. Shader() returns a clip containing commands, and each subsequent to Shader() adds a line to that command chain clip. You must call ExecuteShader() to execute the chain of commands. You must also call ConvertToShader() before and ConvertFromShader() at the end.

The following example will run Diff1 and Diff2 on the clip before returning a Merge of both results. (these shader names are fictive, you have to use real shaders!)

    ConvertToShader(1)
    Input
    Shader("Diff1.cso", Output=2)
    Shader("Diff2.cso", Output=3)
    Shader("Merge.cso", Clip1=2, Clip2=3, Output=1)
    ShaderExecute(last, Input, Clip1Precision=1, Precision=3, OutputPrecision=1)
    ConvertFromShader(1)

It is recommended to use AviSynth MT so that the CPU can work on other threads while waiting for results from the GPU.  
With AviSynth 2.6, ConvertToShader, ConvertFromShader and Shader support MT=1 while ExecuteShader supports MT=2.  
With AviSynth+, all filters run as MT=1 and automatically configure their mode. If you run 8 threads, it will alternate all threads between 2 engines for optimal performance.

#### Shader.avsi functions

#### Common Parameter

MatrixIn/MatrixOut: The input and output color matrix (601, 709, Pc601, Pc709). This can be used for color matrix conversion. Default="709" for both  
FormatOut: The output format. Default = same as input.  
Convert: Whether to call ConvertToShader and ConvertFromShader within the shader. Default=true  
ConvertYuv: Whether do YUV-RGB color conversion. Default=true unless Convert=true and source is RGB  
lsb_in, lsb_out: Whether the input, result of Upscale and output are to be converted to/from DitherTools' Stack16 format. Default=false  
fKernel, fWidth, fHeight, fB, fC: Allows downscaling the output before reading back from GPU. See ResizeShader.  
PlanarIn, PlanarOut: Whether to transfer frame data as 3 individual planes to reduce bandwidth at the expense of extra processing. Generally, PlanarIn brings no performance benefit while PlanarOut brings a nice performance boost. PlanarIn may bring an advantage with larger frames. Default for SuperRes and SuperResXBR: PlanarIn=false, PlanarOut=true. Default for SuperXBR: PlanarIn=true, PlanarOut=true. Default for ResizeShader: PlanarIn=true, PlanarOut=false.  
Engines: In Avisynth+ with MT_NICE_FILTER, sets the number of DirectX engines that will be shared amongst all threads. Set to 2 if running a single shader function for increased performance. Default=1. Ignored in AviSynth 2.6 running with MT_MULTI_INSTANCE.  
Arguments fKernel, fWidth, fHeight, fB, fC are the same as ResizeShader and allows downscaling the output before reading back from GPU  


#### SuperResXBR(Input, Passes, Str, Soft, XbrStr, XbrSharp, Factor, MatrixIn, MatrixOut, FormatOut, Convert, ConvertYuv, lsb_in, lsb_out, fKernel, fWidth, fHeight, fB, fC, PlanarIn, PlanarUpscale, PlanarOut)
Enhances upscaling quality, combining Super-xBR and SuperRes to run in the same command chain, reducing memory transfers and increasing performance.

Arguments Passes, Str, Soft are the same as SuperRes.  
Arguments XbrStr, XbrSharp, Factor are the same as SuperXBR.  


#### SuperRes(Input, Passes, Str, Soft, Upscale, MatrixIn, MatrixOut, FormatOut, Convert, ConvertYuv, lsb_in, lsb_upscale, lsb_out, fKernel, fWidth, fHeight, fB, fC, PlanarIn, PlanarUpscale, PlanarOut)
Enhances upscaling quality.

Arguments:  
Passes: How many SuperRes passes to run. Default=1.  
Str: How agressively we want to run SuperRes, between 0 and 1. Default=1.  
Soft: How much smoothness we want to add, between 0 and 1. Default=0.  
Upscale: An upscaling command that must contain offset-correction. Ex: """nnedi3_rpow2(2, cshift="Spline16Resize")"""  
lsb_upscale: Whether the result of Upscale is to be converted from DitherTools' Stack16 format. Default=false  
PlanarUpscale: Whether to read the result of Upscale as planar data. Default=true (slight performance gain)  


#### SuperXBR(Input, Str, Sharp, Factor, MatrixIn, MatrixOut, FormatOut, Convert, lsb_in, lsb_out, fKernel, fWidth, fHeight, fB, fC, PlanarIn, PlanarOut)
Doubles the size of the image. Produces a sharp result, but with severe ringing.

Arguments:  
Str: Value between 0 and 5 specifying the strength. Default=1.  
Sharp: Value between 0 and 1.5 specifying the weight. Default=1.  
Factor: Upscales the image by a factor of 2, 4, 8 or 16 in one call. Default=2.  


#### ResizeShader(Input, Width, Height, Kernel, B, C, MatrixIn, MatrixOut, FormatOut, Convert, ConvertYuv, lsb_in, lsb_out, PlanarIn, PlanarOut)
Downscales the image in high quality.

Arguments:  
Width: The width to resize to.  
Height: The height to resize to.  
Str: The algorithm strength to apply between 0 and 1. Default=.5  
Soft: If true, the result will be softer. Default=false  
Kernel: The resize algorithm to use: SSim, Bicubic or ColorMatrix. ColorMatrix performs matrix conversion without resizing. Default=Bicubic.  
B, C: When using SSim, B sets the Strength (0 to 1, default=.5) and C sets whether to use a soft algorithm (0 or 1, default=0)  
B, C: When using Bicubic, sets the B and C values. Default is B=1/3, C=1/3.  
When used as a downscaler in other functions, default is fB=0, fC=.75 (useful for downscaling)  


Shiandow provides many other HLSL shaders available here that can be integrated into AviSynth.  
https://github.com/zachsaw/MPDN_Extensions/tree/master/Extensions/RenderScripts



#### Shader.dll functions

#### ConvertToShader(Input, Precision, lsb, Planar, Opt)
Converts a clip into a wider frame containing UINT16 or half-float data. Clips must be converted in such a way before running any shader.

16-bit-per-channel half-float data isn't natively supported by AviSynth. It is stored in a RGB32 container with a Width that is twice larger. When using Clip.Width, you must divine by 2 to get the accurate width.

Arguments:  
Precision: 0 to convert to Y8, 1 to convert into BYTE, 2 to convert into UINT16, 3 to convert into half-float. Default=1  
lsb: Whether to convert from DitherTools' Stack16 format. Only YV12 and YV24 are supported. Default=false  
Planar: True to convert into YV24 planar data to reduce memory transers. If you assign such a clip to Clip1, the shader will receive the 3 planes as Clip1, Clip2 and Clip3. Default=false  
Opt: Optimization path. In Avisynth 2.6, 0 for only C++, 1 for SSE2, 2 for AVX(only used with Precision=3), -1 to auto-detect. 
	In Avisynth+, -1 to use the Avisynth+ code path, other values to use legacy code paths. Default=-1
     

#### ConvertFromShader(Input, Precision, Format, lsb)
Convert a half-float clip into a standard clip.

Arguments:  
Precision: 0 to convert int Y8, 1 to convert into BYTE, 2 to convert into UINT16, 3 to convert into half-float. Default=1  
Format: The video format to convert to. Valid formats are YV12, YV24 and RGB32. Default=YV12.  
lsb: Whether to convert to DitherTools' Stack16 format. Only YV12 and YV24 are supported. Default=false  
Opt: Optimization path. In Avisynth 2.6, 0 for only C++, 1 for SSE2, 2 for AVX(only used with Precision=3), -1 to auto-detect. 
	In Avisynth+, -1 to use the Avisynth+ code path, other values to use legacy code paths. Default=-1

#### Shader(Input, Path, EntryPoint, ShaderModel, Param1-Param9, Clip1-Clip9, Output, Width, Height, Precision, Defines)
Runs a HLSL pixel shader on specified clip. You can either run a compiled .cso file or compile a .hlsl file.

Arguments:  
Input: The first input clip.  
Path: The path to the HLSL pixel shader file to run. If not specified, Clip1 will be copied to Output.  
EntryPoint: If compiling HLSL source code, specified the code entry point.  
ShaderModel: If compiling HLSL source code, specified the shader model. Usually PS_2_0 or PS_3_0  
Param0-Param8: Sets each of the shader parameters.  
Ex: "float4 p4 : register(c4);" will be set with Param4="1,1,1,1f"  
End each value with 'f'(float), 'i'(int) or 'b'(bool) to specify its type.  
Param0 corresponds to c0, Param1 corresponds to c1, etc.  
If setting float or int, you can set a vector or 2, 3 or 4 elements by separating the values with ','.  
If not specified, Param0 = Width,Height and Param1 = 1/Width, 1/Height by default.  
Clip1-Clip9: The index of the clips to set into this shader. Input clips are defined when calling ExecuteShader. Clip1 sets 's0' within the shader, while clip2-clip9 set 's1' to 's8'. The order is important.  
Default for clip1 is 1, for clip2-clip9 is 0 which means no source clip.  
Output: The clip index where to write the output of this shader, between 1 and 9. Default is 1 which means it will be the output of ExecuteShader. If set to another value, you can use it as the input of another shader. The last shader in the chain must have output=1.  
Width, Height: The size of the output texture. Default = same as input texture.  
Precision: While processing precision is set with ExecuteShader, this allows processing certain shaders with a different precision.
Defines: List of pre-compilation constants to set for HLSL files, separated by ';'. Ex: "Kb=0.114;Kr=0.299;"


ChangeLog.txt
Version 1.6.6: June 17th 2020
- Updated headers to support Avisynth+ 3.6

Version 1.6.5: May 13th 2018
- Fixed crash on NVidia cards when PlanarOut=true

Version 1.6.4: September 20th 2017
- Fixed SuperRes when using high-bit-depth upscale

Version 1.6.3: September 8th 2017
- Fixed compatibility regression with Avisynth 2.6
- Now converts properly from YV16 format with Avisynth 2.6
- Updated Avisynth headers

Version 1.6.2: August 9th 2017
- Added Factor parameter to SuperXBR and SuperResXBR to upscale by a factor of 2, 4, 8 or 16 in one call

Version 1.6.1: March 21st 2017
- Fixed color shift with RGB32 input and RGB48 output
- Compiled with Visual Studio 2017

Version 1.6: February 13th 2017
- Added support for all new Avisynth+ clip formats
- Added support for native high-bit-depth clip formats
- ConvertToShader and ConvertFromShader now use native Avisynth+ functions for conversion, which results in better performance in several cases
- AviSynth 2.6 is still supported. ConvertToShader and ConvertFromShader have 2 code paths. Set Opt=2 to use the legacy code path. Opt=-1 will use Avisynth+ functions.
- Updated SSIM downscaler: more advanced XYZ processing
- Fixed bug with ResizeShader and RGB clips

Version 1.5.6: October 19th 2016
- Changed planar format back to D3DFMT_L8
- PlanarIn default value is now false for all methods
- Detect whether graphic card supports planar output and gracefully disable PlanarOut if it doesn't

Version 1.5.5: August 18th 2016
- Updated Avisynth headers
- Moved lock so that it doesn't cause all other filters to run as MT_SERIALIZED
- MT mode now set using SetCacheHints instead of using env2
- Planar format for precision 1 changed from D3DFMT_L8 to D3DFMT_A8 to see if it works better on NVidia cards

Version 1.5.4: July 15th 2016
- MatrixIn/MatrixOut: Changed 601 to Rec601, 709 to Rec709.
- Fixed: It wasn't working on Windows 7
- All HLSL files now pre-compiled and included in DLL
- Much faster initialization
- ExecuteShader: Added Resource parameter. When true, it will only open internal resources without searching for a physical file

Version 1.5.3: July 12th 2016
- Fixed: Filter was accidentally being registered as MT_MULTI_INSTANCE on AviSynth+
- Fixed: Filter did not work with MT_NICE_FILTER
- Fixed: Filter wasn't creating the second core (back to square one with MT_MULTI_INSTANCE)
- ExecuteShader: Added Engines parameter to specify the number of cores to run when using MT_NICE_FILTER
- MT_NICE_FILTER now working as expected
- Performance boost and lower memory usage

Version 1.5.2: July 6th 2016
- All standard shader files are now included within the DLL.
- If a shader file of the same name is also found in the DLL's folder, that one will be taken first.
- Recompiled with AviSynth+ r2022's headers. It won't won't work with older AviSynth+ (AviSynth 2.6 is OK)

Version 1.5.1: June 29th 2016
- Shader: Added Defines parameter to configure HLSL shaders before compilation
- Restructured the files to compile several HLSL files on-demand with compilation parameters. Reduced the amount of shader files from 41 to 26.
- Fixed: Texture overflow method wasn't properly reset to CLAMP after dithering with WRAP
- ExecuteShader: Changed default values Clip1Precision=1, OutputPrecision=1, Precision=3
- ResizeShader: Changed default values to Kernel="Bicubic", B=1./3, C=1./3. Defaults for downscalers remain B=0, C=.75
- MatrixIn and MatrixOut now support PC colors for matrix conversion. Valid values are now 601, 709, Pc601 and Pc709.
- ResizeShader: Added Kernel="ColorMatrix" to perform color matrix conversion without resizing

Version 1.5: June 21th 2016
- Major performance improvements (~double)
- Replaced D3DPRESENT_INTERVAL_DEFAULT with D3DPRESENT_INTERVAL_IMMEDIATE which removed the performance cap
- Added Ordered Dithering (16x16) before returning 8-bit-depth data
- Shader: Clip1-Clip9 can now have values between 0 and 100
- Shader: Added Precision parameter to specify different output precisions per shader
- Shader: Now detecting whether to compile file or not based on extension (.hlsl or .cso)
- Shader: ShaderModel default value is now 'ps_3_0'
- ExecuteShader: Full support for MT_NICE_FILTER under AviSynth+. When MT is enabled, it will alternate threads between 2 engines for optimal performance. AviSynth 2.6 will use it with MT=2.
- ExecuteShader: Added PlanarOut parameter to return output as YV24 planar data
- ExecuteShader: Can now transfer YV24 source clips as planar data
- ExecuteShader: Precision can now be 0 to process with Y8
- ConvertToShader/ConvertFromShader: Added assembly optimizations (thanks to Chikuzen!!)

Src/Init.cpp
#include <algorithm>
#include "ConvertShader.h"
#include "Shader.h"
#include "ExecuteShader.h"
#include "PixelFormatParser.h"
#include "ConvertStacked.hpp"

const int DefaultConvertYuv = false;
static PixelFormatParser pixelFormatParser;

AVSValue __cdecl Create_ConvertToShader(AVSValue args, void* user_data, IScriptEnvironment* env) {
	PClip input = args[0].AsClip();
	bool planar = args[3].AsBool(false);

	const VideoInfo& vi = input->GetVideoInfo();
	if (!vi.IsY() && !vi.Is420() && !vi.Is422() && !vi.Is444() && !vi.IsRGB())
		env->ThrowError("ConvertToShader: Source format is not supported.");

	bool stack16 = args[2].AsBool(false);
	bool HBD = vi.BitsPerComponent() > 8 || stack16; // High-bit-depth source
	int precision = args[1].AsInt(2);
	if (precision < 0 || precision > 3)
		env->ThrowError("ConvertToShader: Precision must be 0, 1, 2 or 3");
	if (stack16 && vi.BitsPerComponent() > 8)
		env->ThrowError("ConvertToShader: Can't use both lsb and high-bit-depth sources!");
	if (stack16 && precision == 1)
		env->ThrowError("ConvertToShader: When using lsb, don't set precision=1!");
	if (stack16 && vi.IsRGB())
		env->ThrowError("ConvertToShader: Conversion from Stack16 only supports YV12 and YV24");
	if (HBD && precision < 2)
		env->ThrowError("ConvertToShader: Precision must be 2 when using high-bit-depth video");

	if (precision == 0) {
		if (!vi.IsY8())
			input = env->Invoke("ConvertToY8", input).AsClip();
		return input;
	}

	int Opt = args[4].AsInt(-1);
	bool isPlusMt = env->FunctionExists("SetFilterMTMode");
	if (!isPlusMt || Opt > -1) {
		// This code is designed for Avisynth 2.6
		if (precision == 1 && planar && vi.IsYUV()) {
			if (!vi.IsYV24()) {
				AVSValue sargs[2] = { input, "Spline36" };
				const char *nargs[2] = { 0, "chromaresample" };
				input = env->Invoke("ConvertToYV24", AVSValue(sargs, 2), nargs).AsClip();
			}
		} else {
			if (vi.IsY8() || vi.IsYV12() || vi.IsYV16()) {
				if (stack16) {
					if (!env->FunctionExists("Dither_resize16nr"))
						env->ThrowError("ConvertToShader: Dither_resize16nr is missing.");
					AVSValue sargs[5] = { input, vi.width, vi.height / 2, "Spline36", "YV24" };
					const char *nargs[5] = { 0, 0, 0, "kernel", "csp" };
					input = env->Invoke("Dither_resize16nr", AVSValue(sargs, 5), nargs).AsClip();
				} else {
					AVSValue sargs[2] = { input, "Spline36" };
					const char *nargs[2] = { 0, "chromaresample" };
					input = env->Invoke("ConvertToYV24", AVSValue(sargs, 2), nargs).AsClip();
				}
			}

			input = new ConvertShader(
				input,					// source clip
				precision,				// precision, 1 for RGB32, 2 for UINT16 and 3 for half-float data.
				stack16,				// lsb / Stack16
				std::string(""),
				planar,					// Planar
				Opt,					// 0 for C++ only, 1 for use SSE2, 2 for use SSSE3 and others for use F16C. -1 to use Avisynth+ functions.
				env);					// env is the link to essential informations, always provide it
		}
	} else {
		if (stack16)
			input = env->Invoke("ConvertFromStacked", input).AsClip();
		else if (precision > 1 && vi.BitsPerComponent() < 16) {
			AVSValue sargs[2] = { input, 16 };
			const char *nargs[2] = { 0, 0 };
			input = env->Invoke("ConvertBits", AVSValue(sargs, 2), nargs).AsClip();
		}

		// With Avisynth+, we'll rely purely on a chain of built-in functions.
		if (vi.IsYUV() || vi.IsYUVA()) {
			if (vi.Is420() || vi.Is422()) {
				AVSValue sargs[2] = { input, "Spline36" };
				const char *nargs[2] = { 0, "chromaresample" };
				input = env->Invoke("ConvertToYUV444", AVSValue(sargs, 2), nargs).AsClip();
			}
			if (!planar) {
				if (vi.NumComponents() == 3)
					input = env->Invoke("AddAlphaPlane", input).AsClip();
				// Cast YUVA as RGBA.
				std::string shuffleFormat = std::string("RGBAP") + (precision == 1 ? "8" : precision == 4 ? "S" : "16");
				AVSValue sargs[4] = { input, precision == 1 ? "RGBA" : "BGRA", "YUVA", shuffleFormat.c_str() };
				const char *nargs[4] = { 0, "planes", "source_planes", "pixel_type" };
				input = env->Invoke("CombinePlanes", AVSValue(sargs, 4), nargs).AsClip();
				// CombinePlanes causes FlipVertical.
				input = env->Invoke(precision > 1 ? "ConvertToRGB64" : "ConvertToRGB32", input).AsClip();
				input = env->Invoke("FlipVertical", input).AsClip();
			}
		} else {
			if (planar) {
				// Cast YUVA as RGBA.
				if (!vi.IsPlanarRGB())
					input = env->Invoke("ConvertToPlanarRGB", input).AsClip();
				else if (vi.NumComponents() == 4)
					input = env->Invoke("RemoveAlphaPlane", input).AsClip();
				std::string shuffleFormat = std::string("YUV444P") + (precision == 1 ? "8" : precision == 4 ? "S" : "16");
				AVSValue sargs[4] = { input, "YUV", precision == 1 ? "BGR" : "RGB", shuffleFormat.c_str() };
				const char *nargs[4] = { 0, "planes", "source_planes", "pixel_type" };
				input = env->Invoke("CombinePlanes", AVSValue(sargs, 4), nargs).AsClip();
				// CombinePlanes causes FlipVertical.
			} else if (vi.BitsPerComponent() == 8) {
				// 8-bit texture format is BGRA and must be switched to RGBA.
				if (!vi.IsPlanarRGB())
					input = env->Invoke("ConvertToPlanarRGBA", input).AsClip();
				else if (vi.NumComponents() == 3)
					input = env->Invoke("AddAlphaPlane", input).AsClip();
				std::string shuffleFormat = std::string("RGBAP") + (precision == 1 ? "8" : precision == 4 ? "S" : "16");
				AVSValue sargs[4] = { input, "BGRA", "RGBA", shuffleFormat.c_str() };
				const char *nargs[4] = { 0, "planes", "source_planes", "pixel_type" };
				input = env->Invoke("CombinePlanes", AVSValue(sargs, 4), nargs).AsClip();
				input = env->Invoke(HBD ? "ConvertToRGB64" : "ConvertToRGB32", input).AsClip();
			} else if (vi.IsPlanarRGB())
				input = env->Invoke(HBD ? "ConvertToRGB64" : "ConvertToRGB32", input).AsClip();
			else if (vi.NumComponents() == 3)
				input = env->Invoke("AddAlphaPlane", input).AsClip();
			if (!planar)
				input = env->Invoke("FlipVertical", input).AsClip();
		}
		if (precision > 1)
			input = env->Invoke("Shader_ConvertToDoubleWidth", input).AsClip();
	}
	return input;
}

AVSValue __cdecl Create_ConvertFromShader(AVSValue args, void* user_data, IScriptEnvironment* env) {
	PClip input = args[0].AsClip();
	const VideoInfo& viSrc = input->GetVideoInfo();

	if (!viSrc.IsRGB32() && !viSrc.IsYV24() && !viSrc.IsY8())
		env->ThrowError("ConvertFromShader: Source must be RGB32, Planar YV24 or Y8");

	int precision = args[1].AsInt(2);
	if (precision < 0 || precision > 3)
		env->ThrowError("ConvertFromShader: Precision must be 0, 1, 2 or 3");

	auto format = std::string(args[2].AsString("YV24"));
	std::transform(format.begin(), format.end(), format.begin(), toupper); // convert lower to UPPER
	VideoInfo viDst = pixelFormatParser.GetVideoInfo(std::string(format));
	bool HBD = viDst.BitsPerComponent() > 8;
	//if (viDst.pixel_type == 0 || (precision > 0 && format != "YV12" && format != "YV24" && format != "RGB24" && format != "RGB32" && !HBD))
	//	env->ThrowError("ConvertFromShader: Destination format is not supported");

	bool stack16 = args[3].AsBool(false);
	if (stack16 && precision == 1)
		env->ThrowError("ConvertFromShader: When using lsb, don't set precision=1!");
	if (stack16 && HBD)
		env->ThrowError("ConvertFromShader: Can't use both lsb and high-bit-depth sources!");
	if (stack16 && (viDst.IsYV12() || viDst.IsY8()) && !env->FunctionExists("Dither_resize16nr")) 
		env->ThrowError("ConvertFromShader: Dither_resize16nr is missing.");
	if (stack16 && !viDst.IsYV12() && !viDst.IsYV24())
		env->ThrowError("ConvertFromShader: Conversion to Stack16 only supports YV12 and YV24");
	if (HBD && precision < 2)
		env->ThrowError("ConvertFromShader: Precision must be 2 when using high-bit-depth");

	int Opt = args[4].AsInt(-1);
	bool isPlusMt = env->FunctionExists("SetFilterMTMode");
	if (!isPlusMt || Opt > -1) {
		// This code is designed for Avisynth 2.6
		if ((precision == 0 || precision == 1) && (viSrc.IsY() || viSrc.IsYV24()) && (viDst.IsY8() || viDst.IsYV12() || viDst.IsYV16() || viDst.IsYV24())) {
			if (viDst.IsY8() && !viSrc.IsY())
				input = env->Invoke("ConvertToY8", input).AsClip();
			if (viDst.IsYV12())
				input = env->Invoke("ConvertToYV12", input).AsClip();
			if (viDst.IsYV16())
				input = env->Invoke("ConvertToYV24", input).AsClip();
			if (viDst.IsYV24() && !viSrc.IsYV24())
				input = env->Invoke("ConvertToYV24", input).AsClip();
		}
		else {
			input = new ConvertShader(
				input,				// source clip
				precision,			// precision, 1 for RGB32, 2 for UINT16 and 3 for half-float data.
				stack16,			// lsb / Stack16
				format,				// destination format
				false,
				Opt,				// 0 for C++ only, 1 for use SSE2, 2 for use SSSE3 and others for use F16C. -1 to use Avisynth+ functions.
				env);				// env is the link to essential informations, always provide it

			if (viDst.IsY8() || viDst.IsYV12() || viDst.IsYV16()) {
				if (stack16) {
					AVSValue sargs[6] = { input, input->GetVideoInfo().width, input->GetVideoInfo().height / 2, "Spline36", format.c_str(), true };
					const char *nargs[6] = { 0, 0, 0, "kernel", "csp", "invks" };
					input = env->Invoke("Dither_resize16nr", AVSValue(sargs, 6), nargs).AsClip();
				}
				else if (viDst.IsYV12())
					input = env->Invoke("ConvertToYV12", input).AsClip();
				else if (viDst.IsY8())
					input = env->Invoke("ConvertToY8", input).AsClip();
			}
		}
	}
	else {
		// With Avisynth+, we'll rely purely on a chain of built-in functions.
		if (precision > 1)
			input = env->Invoke("Shader_ConvertFromDoubleWidth", input).AsClip();

		if (viSrc.IsYUV()) {
			// Planar output.
			if (viDst.IsRGB()) {
				// Cast YUV as RGB.
				std::string shuffleFormat = std::string("RGBP") + (precision == 1 ? "8" : precision == 4 ? "S" : "16");
				AVSValue sargs[4] = { input, precision == 1 ? "BGR" : "BGR", "YUV", shuffleFormat.c_str() };
				const char *nargs[4] = { 0, "planes", "source_planes", "pixel_type" };
				input = env->Invoke("CombinePlanes", AVSValue(sargs, 4), nargs).AsClip();
				// CombinePlanes causes FlipVertical.
				if (!viDst.IsPlanarRGB() && !viDst.IsPlanarRGBA())
					input = env->Invoke(viDst.IsRGB24() ? "ConvertToRGB24" : viDst.IsRGB32() ? "ConvertToRGB32" : viDst.IsRGB48() ? "ConvertToRGB48" : viDst.IsRGB64() ? "ConvertToRGB64" : "", input).AsClip();
			}
		}
		else {
			// Stacked output
			if (!viDst.IsRGB()) {
				// Cast RGB as YUV.
				input = env->Invoke("ConvertToPlanarRGBA", input).AsClip();
				std::string shuffleFormat = std::string("YUVA444P") + (precision == 1 ? "8" : precision == 4 ? "S" : "16");
				AVSValue sargs[4] = { input, "YUVA", precision == 1 ? "RGBA" : "BGRA", shuffleFormat.c_str() };
				const char *nargs[4] = { 0, "planes", "source_planes", "pixel_type" };
				input = env->Invoke("CombinePlanes", AVSValue(sargs, 4), nargs).AsClip();
				// CombinePlanes causes FlipVertical.
			} else if (!HBD) {
				// Cast BGRA as RGBA
				input = env->Invoke("ConvertToPlanarRGBA", input).AsClip();
				std::string shuffleFormat = std::string("RGBAP") + (precision == 1 ? "8" : precision == 4 ? "S" : "16");
				AVSValue sargs[4] = { input, "RGBA", "BGRA", shuffleFormat.c_str() };
				const char *nargs[4] = { 0, "planes", "source_planes", "pixel_type" };
				input = env->Invoke("CombinePlanes", AVSValue(sargs, 4), nargs).AsClip();
				if (!viDst.IsPlanarRGB() && !viDst.IsPlanarRGBA())
					input = env->Invoke(viDst.IsRGB24() ? "ConvertToRGB24" : viDst.IsRGB32() ? "ConvertToRGB32" : viDst.IsRGB48() ? "ConvertToRGB48" : viDst.IsRGB64() ? "ConvertToRGB64" : "", input).AsClip();
			}
			input = env->Invoke("FlipVertical", input).AsClip();
		}

		if (viDst.NumComponents() == 3)
			input = env->Invoke("RemoveAlphaPlane", input).AsClip();
		if (viDst.Is420() || viDst.Is422()) {
			AVSValue sargs[2] = { input, "Spline36" };
			const char *nargs[2] = { 0, "chromaresample" };
			input = env->Invoke(viDst.Is422() ? "ConvertToYUV422" : "ConvertToYUV420", AVSValue(sargs, 2), nargs).AsClip();
		}

		if (precision > 1 && viDst.BitsPerComponent() < 16 && !stack16) {
			if (viDst.BitsPerComponent() == 8) {
				// Dither
				AVSValue sargs[3] = { input, viDst.BitsPerComponent(), -1 };
				const char *nargs[3] = { 0, 0, "dither" };
				input = env->Invoke("ConvertBits", AVSValue(sargs, 3), nargs).AsClip();
			}
			else {
				// 10-14bit: don't dither. "dither=-1" can't be used due to a bug.
				AVSValue sargs[2] = { input, viDst.BitsPerComponent() };
				const char *nargs[2] = { 0, 0 };
				input = env->Invoke("ConvertBits", AVSValue(sargs, 2), nargs).AsClip();
			}
		}

		if (stack16)
			input = env->Invoke("ConvertToStacked", input).AsClip();
	}
	return input;
}

AVSValue __cdecl Create_Shader(AVSValue args, void* user_data, IScriptEnvironment* env) {
	return new Shader(
		args[0].AsClip(),			// source clip
		args[1].AsString(""),		// shader path
		args[2].AsString("main"),	// entry point
		args[3].AsString("ps_3_0"),	// shader model
		args[4].AsString(""),		// param 0
		args[5].AsString(""),		// param 1
		args[6].AsString(""),		// param 2
		args[7].AsString(""),		// param 3
		args[8].AsString(""),		// param 4
		args[9].AsString(""),		// param 5
		args[10].AsString(""),		// param 6
		args[11].AsString(""),		// param 7
		args[12].AsString(""),		// param 8
		args[13].AsInt(1),			// clip 1
		args[14].AsInt(0),			// clip 2
		args[15].AsInt(0),			// clip 3
		args[16].AsInt(0),			// clip 4
		args[17].AsInt(0),			// clip 5
		args[18].AsInt(0),			// clip 6
		args[19].AsInt(0),			// clip 7
		args[20].AsInt(0),			// clip 8
		args[21].AsInt(0),			// clip 9
		args[22].AsInt(1),			// output clip
		args[23].AsInt(0),			// width
		args[24].AsInt(0),			// height
		args[25].AsInt(-1),			// precision
		args[26].AsString(""),		// defines
		env);						// env is the link to essential informations, always provide it
}

AVSValue __cdecl Create_ExecuteShader(AVSValue args, void* user_data, IScriptEnvironment* env) {
	int ParamClipPrecision[9];
	int CurrentPrecision = 1;
	for (int i = 0; i < 9; i++) {
		CurrentPrecision = args[i + 10].AsInt(CurrentPrecision);
		ParamClipPrecision[i] = CurrentPrecision;
	}

	// P.F. IsClip() pre-check, because avisynth debug version give assert if AsClip() is not a clip
	// preventing happy debugging
	return new ExecuteShader(
		args[0].AsClip(),			// source clip containing commands
		args[1].IsClip() ? args[1].AsClip() : nullptr,			// Clip1
		args[2].IsClip() ? args[2].AsClip() : nullptr,			// Clip2
		args[3].IsClip() ? args[3].AsClip() : nullptr,			// Clip3
		args[4].IsClip() ? args[4].AsClip() : nullptr,			// Clip4
		args[5].IsClip() ? args[5].AsClip() : nullptr,			// Clip5
		args[6].IsClip() ? args[6].AsClip() : nullptr,			// Clip6
		args[7].IsClip() ? args[7].AsClip() : nullptr,			// Clip7
		args[8].IsClip() ? args[8].AsClip() : nullptr,			// Clip8
		args[9].IsClip() ? args[9].AsClip() : nullptr,			// Clip9
		ParamClipPrecision,			// ClipPrecision, args[10-18]
		args[19].AsInt(3),			// Precision
		args[20].AsInt(1),			// PrecisionOut
		args[21].AsBool(false),		// PlanarOut
		args[22].AsInt(1),			// Engines count
		args[23].AsBool(false),		// Resource (don't search for file)
		env);
}

AVSValue __cdecl Create_GetBitDepth(AVSValue args, void* user_data, IScriptEnvironment* env) {
	VideoInfo vi;
	AVSValue Format = args[1].AsString("");
	if (Format.AsString() == "")
		vi = args[0].AsClip()->GetVideoInfo();
	else {
		int pixelFormat = pixelFormatParser.GetPixelFormatAsInt(std::string(Format.AsString()));
		if (pixelFormat != 0)
			vi = pixelFormatParser.GetVideoInfo(pixelFormat);
		else
			env->ThrowError("Shader_GetBitDepth: Specified format is invalid");
	}
	return AVSValue(vi.BitsPerComponent());
}

const AVS_Linkage *AVS_linkage = 0;

extern "C" __declspec(dllexport) const char* __stdcall AvisynthPluginInit3(IScriptEnvironment* env, const AVS_Linkage* const vectors) {
	AVS_linkage = vectors;
	env->AddFunction("ConvertToShader", "c[Precision]i[lsb]b[planar]b[opt]i", Create_ConvertToShader, 0);
	env->AddFunction("ConvertFromShader", "c[Precision]i[Format]s[lsb]b[opt]i", Create_ConvertFromShader, 0);
	env->AddFunction("Shader", "c[Path]s[EntryPoint]s[ShaderModel]s[Param0]s[Param1]s[Param2]s[Param3]s[Param4]s[Param5]s[Param6]s[Param7]s[Param8]s[Clip1]i[Clip2]i[Clip3]i[Clip4]i[Clip5]i[Clip6]i[Clip7]i[Clip8]i[Clip9]i[Output]i[Width]i[Height]i[Precision]i[Defines]s", Create_Shader, 0);
	env->AddFunction("ExecuteShader", "c[Clip1]c[Clip2]c[Clip3]c[Clip4]c[Clip5]c[Clip6]c[Clip7]c[Clip8]c[Clip9]c[Clip1Precision]i[Clip2Precision]i[Clip3Precision]i[Clip4Precision]i[Clip5Precision]i[Clip6Precision]i[Clip7Precision]i[Clip8Precision]i[Clip9Precision]i[Precision]i[OutputPrecision]i[PlanarOut]b[Engines]i[Resource]b", Create_ExecuteShader, 0);
	env->AddFunction("Shader_GetBitDepth", "c[format]s", Create_GetBitDepth, 0);

	env->AddFunction("Shader_ConvertFromStacked", "c[bits]i", ConvertFromStacked::Create, 0);
	env->AddFunction("Shader_ConvertToStacked", "c", ConvertToStacked::Create, 0);
	env->AddFunction("Shader_ConvertFromDoubleWidth", "c[bits]i", ConvertFromDoubleWidth::Create, 0);
	env->AddFunction("Shader_ConvertToDoubleWidth", "c", ConvertToDoubleWidth::Create, 0);
	return "Shader plugin";
}
